aimon/reprompting_api/utils.py

```diff
@@ -0,0 +1,183 @@
+                    "instruction": inst.get("instruction", ""),
+                    "score": inst.get("follow_probability", 0.0),
+                    "explanation": inst.get("explanation", "")
+                })
+    # Sort by score (most confident first)
+    failed.sort(key=lambda x: x["score"], reverse=True)
+    return failed
+
+def get_failed_instructions_count(result) -> int:
+    """
+    Count all failed instructions across adherence, groundedness, and toxicity.
+
+    Args:
+        result: AIMon detection result containing `instruction_adherence`, `groundedness`, and `toxicity` sections.
+
+    Returns:
+        int: Total number of failed instructions.
+    """
+    count = 0
+    # Instruction adherence
+    for inst in result.detect_response.instruction_adherence.get("instructions_list", []):
+        if not inst.get("label", True):
+            count += 1
+    # Groundedness
+    for inst in result.detect_response.groundedness.get("instructions_list", []):
+        if not inst.get("label", True):
+            count += 1
+    count += _count_toxicity_failures(result)  # Toxicity
+    return count
+
+def get_residual_error_score(result):
+    """
+    Compute a normalized residual error score (0–1) based on:
+    - Groundedness follow probabilities
+    - Instruction adherence follow probabilities
+    - Toxicity (inverted: 1 - follow_probability)
+
+    Logic:
+    1. Collect follow probabilities for groundedness & adherence.
+    2. For toxicity, use 1 - follow_probability (since high follow = low error).
+    3. Compute a penalized average using the helper.
+    4. Clamp the final score to [0,1].
+    """
+    combined_probs = []
+
+    for source in ["groundedness", "instruction_adherence"]:
+        combined_probs.extend([
+            item["follow_probability"]
+            for item in getattr(result.detect_response, source, {}).get("instructions_list", [])
+        ])
+
+    # For toxicity, invert the follow probability
+    combined_probs.extend([
+        1 - item["follow_probability"]
+        for item in getattr(result.detect_response, "toxicity", {}).get("instructions_list", [])
+    ])
+
+    residual_error_score = penalized_average(combined_probs) if combined_probs else 0.0
+    residual_error_score = min(1.0, max(0.0, residual_error_score))
+    return round(residual_error_score, 2)
+
+
+def penalized_average(probs: List[float]) -> float:
+    """
+    Compute a penalized average of follow probabilities.
+
+    Penalizes probabilities <0.5 more heavily by doubling their penalty. 
+    Probabilities > 0.5 (passed instructions) recieve no penalty
+
+    Args:
+        probs (List[float]): A list of follow probabilities.
+
+    Returns:
+        float: Penalized average.
+    """
+    penalties = []
+    for p in probs:
+        if p >= 0.5:
+            penalty = 0
+        else:
+            penalty = (1 - p) * 2  # heavier penalty
+        penalties.append(penalty)
+    return sum(penalties) / len(penalties)
```

Comment:
since this function is in utils, I would recommend handling the "divide by zero" error in case someone calls the `penalized_average` function with an empty `probs` list